src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -537,70 +547,136 @@ public static class SystemNamespaceType
                     }
                 }
 
-                return LanguageConstants.String;
+                return new(LanguageConstants.String);
             };
 
-        private static TypeSymbol LoadTextContentTypeBuilder(IBinder binder, IFileResolver fileResolver, IDiagnosticWriter diagnostics, ImmutableArray<FunctionArgumentSyntax> arguments, ImmutableArray<TypeSymbol> argumentTypes)
+        private static FunctionResult LoadTextContentResultBuilder(IBinder binder, IFileResolver fileResolver, IDiagnosticWriter diagnostics, ImmutableArray<FunctionArgumentSyntax> arguments, ImmutableArray<TypeSymbol> argumentTypes)
         {
-            if (argumentTypes[0] is not StringLiteralType filePathType)
+            return TryLoadTextContentFromFile(binder, fileResolver, diagnostics,
+                (arguments[0], argumentTypes[0]),
+                arguments.Length > 1 ? (arguments[1], argumentTypes[1]) : null,
+                out var fileContent,
+                LanguageConstants.MaxLiteralCharacterLimit)
+                ? new(new StringLiteralType(fileContent))
+                : new(LanguageConstants.String);
+        }
+
+        private static FunctionResult LoadJsonContentResultBuilder(IBinder binder, IFileResolver fileResolver, IDiagnosticWriter diagnostics, ImmutableArray<FunctionArgumentSyntax> arguments, ImmutableArray<TypeSymbol> argumentTypes)
+        {
+            string? tokenSelectorPath = null;
+            if (arguments.Length > 1)
             {
-                diagnostics.Write(DiagnosticBuilder.ForPosition(arguments[0]).CompileTimeConstantRequired());
-                return LanguageConstants.String;
+                if (argumentTypes[1] is not StringLiteralType tokenSelectorType)
+                {
+                    diagnostics.Write(DiagnosticBuilder.ForPosition(arguments[1]).CompileTimeConstantRequired());
+                    return new(LanguageConstants.Any);
+                }
+                tokenSelectorPath = tokenSelectorType.RawStringValue;
+            }
+            if (!TryLoadTextContentFromFile(binder, fileResolver, diagnostics,
+                    (arguments[0], argumentTypes[0]),
+                    arguments.Length > 2 ? (arguments[2], argumentTypes[2]) : null,
+                    out var fileContent,
+                    LanguageConstants.MaxJsonFileCharacterLimit))
```

Comment:
Although both approaches are a bit hand-wavey, wouldn't it be more accurate to estimate the size of the JSON after doing the jPath-filtering?

I'm thinking about scenarios where the user has a huge JSON file but only wants to select a small section of it - feels like it would be useful to support this.

Reply:
I wanted to avoid reading huge files, but perhaps we can have some big limit and do exact check after. however I feel we could think about it once people start hitting it?

Reply:
Sure, sounds good to me! Changing this in future won't be a breaking change, so I'm fine waiting to see if we get any requests to change it.